### PR TITLE
Convert all of the launcher crates to use prost.

### PR DESCRIPTION
**Pull request by raskchanky**
This is the last of the crates in the habitat repo to be converted from
protobuf to prost.  Note that unlike the supervisor and
butterfly, I decided against including the generated code in the repo.

It's also worth noting that unlike our approach with the protobuf crate,
we're not actually using the structs generated by prost except at the
edges. Prost rightfully uses Options to represent optional fields and
since all of our protobuf messages have all optional fields, the
ergonomics of using Prost generated structs is sub-optimal. For this
reason, we have our own set of internal structs that mirror the Prost
ones that we use in the launcher and launcher-client crates, and
launcher-protocol is the only crate that knows about the Prost generated
structs.

Closes https://github.com/habitat-sh/habitat/issues/6073

Signed-off-by: Josh Black <raskchanky@gmail.com>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/launcher-protocol/src/types.rs [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+}
+
+pub trait FromProto<T: prost::Message + Default>: Sized {
+    fn from_proto(value: T) -> Result<Self>;
+}
+
+pub trait LauncherMessage: MessageStatic + Clone + Default + fmt::Debug {
+    fn to_bytes(&self) -> Result<Vec<u8>>;
+    fn from_bytes(bytes: &[u8]) -> Result<Self>;
+}
```
Comment (baumanj) [inline components/launcher-protocol/src/types.rs]: Here's an alternate approach that doesn't need macros:

We can consolidate the `MessageStatic`, and `FromProto` traits into `LauncherMessage`, and by adding an associated type to hold the Prost-generated struct, we get:
```rust
pub trait LauncherMessage
where
    Self: Clone + fmt::Debug,
{
    type Generated: prost::Message + Default + From<Self>;
    const MESSAGE_ID: &'static str;

    fn from_proto(value: Self::Generated) -> Result<Self>;

    fn from_bytes(bytes: &[u8]) -> Result<Self> {
        let decoded = Self::Generated::decode(bytes)?;
        Self::from_proto(decoded)
    }

    fn to_bytes(&self) -> Result<Vec<u8>> {
        let envelope = Self::Generated::from(self.clone());
        let mut buf = bytes::BytesMut::with_capacity(envelope.encoded_len());
        envelope.encode(&mut buf)?;
        Ok(buf.to_vec())
    }
}
```

The `impl` changes follow from that, but [here's a full diff](https://github.com/habitat-sh/habitat/files/2818106/associated-type-refactor.diff.txt) if you want to go that way.

Comment (raskchanky) [reply components/launcher-protocol/src/types.rs]: This is exactly the solution I was trying to find from the beginning.  Thank you so much for this.

File: components/launcher-client/src/client.rs [hunk 2]
```diff
@@ -12,17 +12,13 @@
...
-use std::io;
-use std::path::Path;
-
-use crate::core::os::process::Pid;
-use crate::protocol;
+use crate::{
+    core::os::process::Pid,
+    error::{Error, Result},
+    protocol::{self, Error as ProtocolError},
+};
```
[the remark this answers is not in the document]
Comment (raskchanky) [reply components/launcher-client/src/client.rs]: Gotcha, will update

File: components/launcher-client/src/client.rs [hunk 3]
```diff
@@ -164,34 +158,35 @@ impl LauncherCli {
...
         }
         if let Some(gid) = group_id {
-            msg.set_svc_group_id(gid);
+            msg.svc_group_id = Some(gid);
         }
 
         // This is only for Windows
         if let Some(password) = password {
-            msg.set_svc_password(password.to_string());
+            msg.svc_password = Some(password.to_string());
```
[the remark this answers is not in the document]
Comment (raskchanky) [reply components/launcher-client/src/client.rs]: That's awesome, will do.

[the remark this answers is not in the document]
Comment (christophermaier) [reply components/launcher-client/src/client.rs]: @raskchanky Actually, if you wanna get _really_ wild, try this:

```rust
        let msg = protocol::Spawn {
            binary: bin.as_ref().to_path_buf().to_string_lossy().into_owned(),
            svc_user: user.map(|u| u.to_string()),
            svc_group: group.map(|g| g.to_string()),
            svc_user_id: user_id,
            svc_group_id: group_id,
            svc_password: password.map(|p| p.to_string()),
            env: env,
            id: id.to_string(),
            ..Default::default()
        };
```

[the remark this answers is not in the document]
Comment (raskchanky) [reply components/launcher-client/src/client.rs]: 😲

[the remark this answers is not in the document]
Comment (christophermaier) [reply components/launcher-client/src/client.rs]: This pattern could be used on all these messages, actually. Kinda nice to reduce the mutability of everything to the bare minimum.

**Comment by christophermaier** [comment]
@raskchanky Looks great! Once the release is out, let's merge this!

Appveyor timed out (sigh)... so I restarted it.

**Pull request merged**